### PR TITLE
feat: Robust surface matching for Geant4

### PR DESCRIPTION
**Pull request by asalzburger**
This PR fixes:

https://github.com/acts-project/acts/issues/3165

It first creates a `std::multimap`  that relates the `G4VPhysVolume` to the possible `Acts::Surface` objects (there can be more due to the replica mechanism).

In the Stepping action, the right one will then be picked.

The output performance change in geant_hits is expected.

Digitisation configuration file:
[odd-digi-geometric-config.json](https://github.com/acts-project/acts/files/15188041/odd-digi-geometric-config.json)

Example how to run full simulation with this:
```sh
python <path_to_your_acts_source_dir>Examples/Scripts/Python/sim_digi_odd.py --digi-config odd-digi-geometric-config.json --geant4 --gun-multiplicity 10 --gun-pt-range 100 100 -n100 -o g4_mu_100GeV_m10
```

Options are:
```sh
usage: sim_digi_odd.py [-h] [--output OUTPUT] [--events EVENTS] [--skip SKIP] [--edm4hep EDM4HEP] [--geant4] [--ttbar] [--ttbar-pu TTBAR_PU] [--gun-multiplicity GUN_MULTIPLICITY] [--gun-eta-range GUN_ETA_RANGE [GUN_ETA_RANGE ...]] [--gun-pt-range GUN_PT_RANGE [GUN_PT_RANGE ...]] [--rnd-seed RND_SEED]
                       [--digi-config DIGI_CONFIG]

Sim-digi chain with the OpenDataDetector

options:
  -h, --help            show this help message and exit
  --output OUTPUT, -o OUTPUT
                        Output directory
  --events EVENTS, -n EVENTS
                        Number of events
  --skip SKIP, -s SKIP  Number of events
  --edm4hep EDM4HEP     Use edm4hep inputs
  --geant4              Use Geant4 instead of fatras
  --ttbar               Use Pythia8 (ttbar, pile-up 200) instead of particle gun
  --ttbar-pu TTBAR_PU   Number of pile-up events for ttbar
  --gun-multiplicity GUN_MULTIPLICITY
                        Multiplicity of the particle gun
  --gun-eta-range GUN_ETA_RANGE [GUN_ETA_RANGE ...]
                        Eta range of the particle gun
  --gun-pt-range GUN_PT_RANGE [GUN_PT_RANGE ...]
                        Pt range of the particle gun (GeV)
  --rnd-seed RND_SEED   Random seed
  --digi-config DIGI_CONFIG
                        Digitization configuration file
```

### CODE REVIEW

File: Examples/Algorithms/Geant4/src/SensitiveSurfaceMapper.cpp [hunk 1]
```diff
@@ -86,15 +88,18 @@ void ActsExamples::SensitiveSurfaceMapper::remapSensitiveNames(
...
-      mappedVolumeName += std::to_string(mappedSurface->geometryId().value());
       ACTS_VERBOSE("Found matching surface " << mappedSurface->geometryId()
                                              << " at position "
                                              << g4RelPosition.transpose());
-      ACTS_VERBOSE("Remap: " << g4PhysicalVolume->GetName() << " -> "
-                             << mappedVolumeName);
-      g4PhysicalVolume->SetName(mappedVolumeName.c_str());
+      // Check if the prefix is not yet assigned
+      std::string prefixName = std::string(mappingPrefix);
+      if (volumeName.find(prefixName) == std::string::npos) {
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/SensitiveSurfaceMapper.cpp:96]:
```suggestion
      if (volumeName.find(mappingPrefix) == std::string::npos) {
```

File: Examples/Algorithms/Geant4/src/SensitiveSurfaceMapper.cpp [hunk 2]
```diff
@@ -86,15 +88,18 @@ void ActsExamples::SensitiveSurfaceMapper::remapSensitiveNames(
...
+        // Set the new name
+        std::string mappedName = prefixName + volumeName;
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/SensitiveSurfaceMapper.cpp:98]:
```suggestion
        std::string mappedName = std::string(prefixName) + volumeName;
```

File: Examples/Algorithms/Geant4/src/SensitiveSteppingAction.cpp [hunk 3]
```diff
@@ -125,19 +128,53 @@ void ActsExamples::SensitiveSteppingAction::UserSteppingAction(
...
+  // Find the range of candidate surfaces for the current position in the
+  // mapping multimap
+  auto [bsf, esf] = m_surfaceMapping.equal_range(volume);
+  std::size_t nSurfaces = std::distance(bsf, esf);
+
+  ACTS_VERBOSE("Found " << nSurfaces << " candidate surfaces for volume "
+                        << volume->GetName());
+
+  if (nSurfaces == 0) {
+    ACTS_ERROR("No candidate surfaces found for volume " << volume->GetName());
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/SensitiveSteppingAction.cpp:157]:
```suggestion
    ACTS_ERROR("No candidate surfaces found for volume " << volumeName);
```

File: Examples/Algorithms/Geant4/src/SensitiveSteppingAction.cpp [hunk 4]
```diff
@@ -125,19 +128,53 @@ void ActsExamples::SensitiveSteppingAction::UserSteppingAction(
...
+
+  Acts::GeometryIdentifier geoId{};
+
+  // Find the range of candidate surfaces for the current position in the
+  // mapping multimap
+  auto [bsf, esf] = m_surfaceMapping.equal_range(volume);
+  std::size_t nSurfaces = std::distance(bsf, esf);
+
+  ACTS_VERBOSE("Found " << nSurfaces << " candidate surfaces for volume "
+                        << volume->GetName());
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/SensitiveSteppingAction.cpp:154]:
```suggestion
                        << volumeName);
```

File: Examples/Algorithms/Geant4/src/SensitiveSteppingAction.cpp [hunk 5]
```diff
@@ -125,19 +128,53 @@ void ActsExamples::SensitiveSteppingAction::UserSteppingAction(
   }
 
   // Get the physical volume & check if it has the sensitive string name
-  std::string_view volumeName = track->GetVolume()->GetName();
+  auto volume = track->GetVolume();
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/SensitiveSteppingAction.cpp:131]:
```suggestion
  const G4VPhysicalVolume* volume = track->GetVolume();
```

File: Examples/Algorithms/Geant4/src/SensitiveSteppingAction.cpp [hunk 6]
```diff
@@ -125,19 +128,53 @@ void ActsExamples::SensitiveSteppingAction::UserSteppingAction(
...
+    ACTS_ERROR("No candidate surfaces found for volume " << volume->GetName());
+    return;
+  } else if (nSurfaces == 1u) {
+    geoId = bsf->second->geometryId();
+    ACTS_VERBOSE("Unique assignment successful -> to surface " << geoId);
+  } else {
+    // Find the closest surface to the current position
+    Acts::GeometryContext gctx;
+    for (; bsf != esf; ++bsf) {
+      auto surface = bsf->second;
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/SensitiveSteppingAction.cpp:166]:
```suggestion
      Acts::Surface* surface = bsf->second;
```

File: Examples/Algorithms/Geant4/src/SensitiveSteppingAction.cpp [hunk 7]
```diff
@@ -125,19 +128,53 @@ void ActsExamples::SensitiveSteppingAction::UserSteppingAction(
...
+      auto translation = touchable->GetTranslation();
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/SensitiveSteppingAction.cpp:167]:
```suggestion
      const G4ThreeVector& translation = touchable->GetTranslation();
```

**Review by andiwand — APPROVED** [review]
A couple of suggestions not thing critical

I would like to try see if `G4VPVParameterisation` can be used for ID together with the volume after this is in.

**Review by andiwand — APPROVED** [review]

**Comment by krasznaa** [comment]
:thinking: I ran a set of simulations with this latest version of the code this morning. It seems to mostly work... But there are some worrisome things that I observed.

First off, I ran 25 simulation jobs with the following:

```sh
# Run all the muon simulations.
for NMUON in 1 5 10 50 100
do
    for PT in 1 5 10 50 100
    do
	python3 "${ACTS_DIR}/Examples/Scripts/Python/sim_digi_odd.py" \
		--digi-config odd-digi-geometric-config.json \
		--geant4 \
		--gun-multiplicity ${NMUON} \
		--gun-pt-range ${PT} ${PT} \
		--events 100 \
		--output geant4_${NMUON}muon_${PT}GeV 2>&1 | \
	    tee geant4_${NMUON}muon_${PT}GeV.log
    done
done
```

(This is part of a script that I'll share in the PR collecting the new files into traccc.)

Most of these simulations went through without a visible issue. But some of them produced some humongous log files. :thinking:

```
[bash][celeborn]:acts > ls -lhS geant4_*.log
-rw-r--r-- 1 root root  32M May  3 11:17 geant4_100muon_100GeV.log
-rw-r--r-- 1 root root  28M May  3 11:00 geant4_5muon_100GeV.log
-rw-r--r-- 1 root root  25M May  3 11:08 geant4_50muon_100GeV.log
-rw-r--r-- 1 root root 8.9M May  3 11:06 geant4_50muon_50GeV.log
-rw-r--r-- 1 root root 271K May  3 11:12 geant4_100muon_10GeV.log
-rw-r--r-- 1 root root 271K May  3 11:02 geant4_10muon_100GeV.log
-rw-r--r-- 1 root root 271K May  3 11:09 geant4_100muon_1GeV.log
-rw-r--r-- 1 root root 271K May  3 11:11 geant4_100muon_5GeV.log
-rw-r--r-- 1 root root 271K May  3 11:01 geant4_10muon_10GeV.log
-rw-r--r-- 1 root root 271K May  3 11:02 geant4_10muon_50GeV.log
-rw-r--r-- 1 root root 271K May  3 10:58 geant4_1muon_100GeV.log
-rw-r--r-- 1 root root 271K May  3 11:05 geant4_50muon_10GeV.log
-rw-r--r-- 1 root root 271K May  3 11:14 geant4_100muon_50GeV.log
-rw-r--r-- 1 root root 271K May  3 11:01 geant4_10muon_1GeV.log
-rw-r--r-- 1 root root 271K May  3 11:01 geant4_10muon_5GeV.log
-rw-r--r-- 1 root root 271K May  3 10:58 geant4_1muon_10GeV.log
-rw-r--r-- 1 root root 271K May  3 10:58 geant4_1muon_50GeV.log
-rw-r--r-- 1 root root 271K May  3 11:03 geant4_50muon_1GeV.log
-rw-r--r-- 1 root root 271K May  3 11:04 geant4_50muon_5GeV.log
-rw-r--r-- 1 root root 271K May  3 10:59 geant4_5muon_10GeV.log
-rw-r--r-- 1 root root 271K May  3 11:00 geant4_5muon_50GeV.log
-rw-r--r-- 1 root root 271K May  3 10:57 geant4_1muon_1GeV.log
-rw-r--r-- 1 root root 271K May  3 10:58 geant4_1muon_5GeV.log
-rw-r--r-- 1 root root 271K May  3 10:59 geant4_5muon_1GeV.log
-rw-r--r-- 1 root root 271K May  3 10:59 geant4_5muon_5GeV.log
[bash][celeborn]:acts >
```

With the large sizes coming from a **lot** of warnings like:

```
09:15:25    Geant4Partic   WARNING   Particle ID for track ID 361043 not registered. Skip
09:15:25    Geant4Partic   WARNING   Particle ID for track ID 361044 not registered. Skip
09:15:25    Geant4Partic   WARNING   Particle ID for track ID 361045 not registered. Skip
09:15:25    Geant4Partic   WARNING   Particle ID collision with 33|0|1|12|145 detected for initial particles. Skip particle
09:15:25    Geant4Partic   WARNING   Particle ID for track ID 361042 not registered. Skip
```

Should I be worried about all those warnings?

**Pull request merged**